sympy/matrices/matrices.py

```diff
@@ -4587,3 +4595,105 @@ def a2idx(j, n=None):
         if not (j >= 0 and j < n):
             raise IndexError("Index out of range: a[%s]" % (j, ))
     return int(j)
+
+def _find_reasonable_pivot(col, iszerofunc=_iszero, simpfunc=_simplify):
+    """ Find the lowest index of an item in `col` that is
+    suitable for a pivot.  If `col` consists only of
+    Floats, the pivot with the largest norm is returned.
+    Otherwise, the first element where `iszerofunc` returns
+    False is used.  If `iszerofunc` doesn't return false,
+    items are simplified and retested until a suitable
+    pivot is found.
+
+    Returns a 4-tuple
+        (pivot_offset, pivot_val, assumed_nonzero, newly_determined)
+    where pivot_offset is the index of the pivot, pivot_val is
+    the (possibly simplified) value of the pivot, assumed_nonzero
+    is True if an assumption that the pivot was non-zero
+    was made without being probed, and newly_determined are
+    elements that were simplified during the process of pivot
+    finding."""
+
+    newly_determined = []
+    col = list(col)
+    # a column that contains a mix of floats and integers
+    # but at least one float is considered a numerical
+    # column, and so we do partial pivoting
+    if all(isinstance(x, (Float, Integer)) for x in col) and any(isinstance(x, Float) for x in col):
```

Comment:
If we have exact types, partial pivoting is unneeded.  I allow floats and ints because I figure someone typing `[1.3, 1, 1]` should be treated as if someone typed `[1.3, 1.0, 1.0]`.  If you look at the bugs resulting from not doing partial pivoting, most example matrices they have are a mix of floats and ints.  If someone does `[1.3, 1, S(3)/2]`, I'm not exactly sure what we should assume.  Further, if we have the absence of a float, e.g. `[1, 1, S(3)/2]` we have all exact types.  I think the appropriate thing to do in that case is to keep it exact and not partial pivot.  (Partial pivoting is a prelude to a `FloatMatrix` type where the `iszerofunc` will be automatically based on the condition number.)